packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp

```diff
@@ -8613,13 +8632,49 @@ CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
 #ifdef HAVE_TPETRA_MMM_TIMINGS
       Teuchos::TimeMonitor MMrc(*TimeMonitor::getNewTimer(prefix + std::string("TAFC sortAndMergeCrsEntries")));
 #endif
-      Import_Util::sortAndMergeCrsEntries (CSR_rowptr (),
-                                           CSR_colind_LID (),
-                                           CSR_vals ());
+      // Just to be safe we always static cast the Scalar pointer
+      // to the undelying impl_scalar type.
+      typename row_ptrs_host_view_type::non_const_type            CSR_rowptr_host(CSR_rowptr.getRawPtr(), CSR_rowptr.size());
+      typename nonconst_local_inds_host_view_type::non_const_type CSR_colind_LID_host(CSR_colind_LID.getRawPtr(), CSR_colind_LID.size());
+      typename nonconst_values_host_view_type::non_const_type     CSR_vals_host(reinterpret_cast<impl_scalar_type*>(CSR_vals.getRawPtr()), CSR_vals.size());
+
+      typename row_ptrs_device_view_type::non_const_type   CSR_rowptr_dev("rowmap", CSR_rowptr.size());
+      typename local_inds_device_view_type::non_const_type CSR_colind_LID_dev("colind", CSR_colind_LID.size());
+      typename values_device_view_type::non_const_type     CSR_vals_dev("values", CSR_vals.size());
+
+      Kokkos::deep_copy(CSR_rowptr_dev, CSR_rowptr_host);
+      Kokkos::deep_copy(CSR_colind_LID_dev, CSR_colind_LID_host);
+      Kokkos::deep_copy(CSR_vals_dev, CSR_vals_host);
+
+      Import_Util::sortAndMergeCrsEntries (CSR_rowptr_dev,
+                                           CSR_colind_LID_dev,
+                                           CSR_vals_dev);
+
+      // Resize before deep_copy to have matching extents
+      // between host and device or we will run into trouble...
+      if(CSR_colind_LID_host.extent(0) != CSR_colind_LID_dev.extent(0)) {
+	Kokkos::resize(CSR_colind_LID_host, CSR_colind_LID_dev.extent(0));
+	Kokkos::resize(CSR_vals_host, CSR_vals_dev.extent(0));
+      }
+
+      // Copy back to host so we can later pass it to ArrayRCPs
+      Kokkos::deep_copy(CSR_rowptr_host, CSR_rowptr_dev);
+      Kokkos::deep_copy(CSR_colind_LID_host, CSR_colind_LID_dev);
+      Kokkos::deep_copy(CSR_vals_host, CSR_vals_dev);
+
+      // First resize the ArrayRCPs correctly so meta data
+      // matches actual data and avoid suprizes down the road
       if (CSR_rowptr[N] != mynnz) {
         CSR_colind_LID.resize (CSR_rowptr[N]);
         CSR_vals.resize (CSR_rowptr[N]);
       }
+
+      // Do the actual data copy, after this the matrix local
+      // data should be correctly defined.
+      typename nonconst_local_inds_host_view_type::non_const_type tmp_CSR_colind_LID_host(CSR_colind_LID.getRawPtr(), CSR_colind_LID.size());
+      typename nonconst_values_host_view_type::non_const_type     tmp_CSR_vals_host(reinterpret_cast<impl_scalar_type*>(CSR_vals.getRawPtr()), CSR_vals.size());
+      Kokkos::deep_copy(tmp_CSR_colind_LID_host, CSR_colind_LID_host);
+      Kokkos::deep_copy(tmp_CSR_vals_host, CSR_vals_host);
```

Comment:
```suggestion
      Import_Util::sortAndMergeCrsEntries (CSR_rowptr_dev,
                                           CSR_colind_LID_dev,
                                           CSR_vals_dev);
      size_t mynnzMerged = CSR_colind_LID_dev.extent(0);
      if (mynnz != mynnzMerged) {
        // update rowptr, and resize colind/vals
        Kokkos::deep_copy(CSR_rowptr_host, CSR_rowptr_dev);
        CSR_colind_LID.resize (mynnzMerged);
        CSR_vals.resize (mynnzMerged);
        CSR_colind_LID_host = typename nonconst_local_inds_host_view_type::non_const_type (CSR_colind_LID.getRawPtr(), mynnzMerged);
        CSR_vals_host = typename nonconst_values_host_view_type::non_const_type     (reinterpret_cast<impl_scalar_type*>(CSR_vals.getRawPtr()), mynnzMerged);
      }
      // then copy colinds and vals
      Kokkos::deep_copy(CSR_colind_LID_host, CSR_colind_LID_dev);   
      Kokkos::deep_copy(CSR_vals_host, CSR_vals_dev);
```

Reply:
Does this look reasonable? It avoid the extra host-host copy and resizing of host views.

Reply:
Yeah, that might be a bit better but again we want to get rid of all this code as soon as this merges so... maybe not worth rewriting it at this point?

Reply:
Sure, if it's going to get replaced right away